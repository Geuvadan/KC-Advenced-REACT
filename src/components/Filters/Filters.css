.filters-main {
  min-height: 3rem;
  padding: 0.5rem;
  line-height: 2rem;
  background-color: lightgrey;
  display: grid;
  grid-template-columns: 0.5fr 5fr;
  font-size: 0.8rem;
}

.filters-form {
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}

select {
  height: 2rem;
  background-color: #fff;
  width: 3rem;
}

.btn {
  height: 2rem;
  padding: 0.5rem;
  margin: 0 0.3rem;
  font-size: 0.8rem;
  background: darkgrey;
  cursor: pointer;
  text-decoration: none;
  color: black;
}

.input-name {
  width: 5rem;
  height: 2rem;
  font-size: 1rem;
  margin-left: 0.5rem;
}

.input-price {
  width: 3rem;
  height: 2rem;
  font-size: 1rem;
  margin-left: 0.5rem;
}

.rangePrice {
  width: 4rem;
}

fieldset {
  border: none;
  margin: 0 0.5rem;
}

@media all and (max-width: 1090px) {
  .rangePrice {
    display: none;
  }
}
