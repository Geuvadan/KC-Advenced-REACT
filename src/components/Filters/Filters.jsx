import React from 'react';
import './Filters.css';
import { Context } from '../Adset/Context';

export default class Filters extends React.Component {
  constructor(props) {
    super(props);

    this.state = this.context;
    this.state = {
      query: '',
      inputName: null,
      selectedTag: null,
      selectedType: null,
      selectedPrice: 0,
    };
  }

  selTypeChange = (evt) => {
    console.log(evt.target.value);
    this.context.saveType(evt.target.value);
    this.setState({
      selectedType: evt.target.value,
    });
  };

  selTagChange = (evt) => {
    console.log(evt.target.value);
    this.context.saveTag(evt.target.value);
    this.setState({
      selectedTag: evt.target.value,
    });
  };

  maxPrice = () => {
    const { ads } = this.context;
    const prices = ads.map((ad) => {
      return ad.price;
    });
    return Math.max.apply(Math, prices);
  };

  selPriceRangeChange = (evt) => {
    console.log(evt.target.value);
    document.getElementById('price').value = document.getElementById('priceSelector').value;
    this.context.savePrice(evt.target.value);
    this.setState({
      selectedPrice: evt.target.value,
    });
  };

  inputPriceChange = (evt) => {
    console.log(evt.target.value);
    this.context.savePrice(evt.target.value);
    this.setState({
      selectedPrice: evt.target.value,
    });
  };

  inputNameChange = (evt) => {
    console.log(evt.target.value);
    this.context.saveName(evt.target.value);
    this.setState({
      inputName: evt.target.value,
    });
  };

  queryParams = () => {
    let queryString = '';
    if (this.state.inputName !== null && queryString !== '') {
      queryString += `&name=${this.state.inputName}`;
    } else if (this.state.inputName !== null) {
      queryString += `name=${this.state.inputName}`;
    }

    if (this.state.selectedTag !== null && this.state.selectedTag !== 'all' && queryString !== '') {
      queryString += `&tag=${this.state.selectedTag}`;
    } else if (this.state.selectedTag !== null && this.state.selectedTag !== 'all') {
      queryString += `tag=${this.state.selectedTag}`;
    }
    if (this.state.selectedType === 'buy' && queryString !== '') {
      queryString += `&venta=false`;
    } else if (this.state.selectedType === 'buy') {
      queryString += `venta=false`;
    }
    if (this.state.selectedType === 'sell' && queryString !== '') {
      queryString += `&venta=true`;
    } else if (this.state.selectedType === 'sell') {
      queryString += `venta=true`;
    }
    if (
      this.state.selectedPrice !== undefined &&
      this.state.selectedPrice !== 0 &&
      queryString !== ''
    ) {
      queryString += `&price=1-${this.state.selectedPrice}`;
    } else if (this.state.selectedPrice !== undefined && this.state.selectedPrice !== 0) {
      queryString += `price=1-${this.state.selectedPrice}`;
    }

    return queryString;
  };

  filterBtn = (evt) => {
    evt.preventDefault();
    const params = this.queryParams();
    console.log(params);
    this.context.saveQuery(params);
  };

  resetBtn = () => {
    this.setState({
      query: '',
      inputName: null,
      selectedTag: null,
      selectedType: null,
      selectedPrice: 0,
    });
  };

  render() {
    const { tags } = this.context;
    return (
      <div className="filters-main">
        <div>Filters</div>
        <form className="filters-form" onSubmit={this.filterBtn}>
          <fieldset>
            <label>Nombre: </label>
            <input className="input-name" type="text" onChange={this.inputNameChange} />
          </fieldset>
          <fieldset>
            <label>Operación: </label>
            <select onChange={this.selTypeChange}>
              <option value="all">Todas</option>
              <option value="buy">Compra</option>
              <option value="sell">Venta</option>
            </select>
          </fieldset>
          <fieldset>
            <label>Tag: </label>
            <select onChange={this.selTagChange}>
              <option value="all">Todas</option>
              {tags.map((tag, index) => (
                <option key={index} value={tag}>
                  {tag}
                </option>
              ))}
            </select>
          </fieldset>
          <fieldset>
            <label>Precio máximo: </label>
            <input
              id="priceSelector"
              type="range"
              min="0"
              max={this.maxPrice()}
              onChange={this.selPriceRangeChange}
            />

            <input
              className="input-price"
              type="text"
              id="price"
              onChange={this.inputPriceChange}
            />
          </fieldset>

          <input className="btn" type="submit" />
          <input className="btn" type="reset" onClick={this.resetBtn} />
        </form>
      </div>
    );
  }
}

Filters.contextType = Context;
