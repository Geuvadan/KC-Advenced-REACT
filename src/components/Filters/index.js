import { connect } from 'react-redux';

import Filters from './Filters';

import { setFilter, fetchAds } from '../../store/actions';

function mapStateToProps(state, ownProps) {
  return {
    queryFilter: state.queryFilter,
    tags: state.tagsAvailable,
    ads: state.ads,
  };
}

const mapDispatchToProps = {
  setFilter,
  fetchAds,
};

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
