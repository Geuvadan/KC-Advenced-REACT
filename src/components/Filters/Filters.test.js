import React from 'react';
import { shallow } from 'enzyme';
import Filters from './Filters';

describe('Filters', () => {
  const props = {
    ads: [
      {
        _id: '1234567890',
        name: 'test_Ad_1',
        type: 'sell',
        price: 100,
      },
      {
        _id: '1234567890',
        name: 'test_Ad_1',
        type: 'sell',
        price: 100,
      },
    ],
    fetchAds: jest.fn(),
    setFilter: jest.fn(),
    tags: ['tag1', 'tag2', 'tag3'],
  };

  const wrapper = shallow(<Filters {...props} />);

  test('Snapshot testing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Shoud fetchAds on form submit', () => {
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(props.fetchAds).toHaveBeenCalled();
  });
});
