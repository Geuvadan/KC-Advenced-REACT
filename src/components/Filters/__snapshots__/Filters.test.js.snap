// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Snapshot testing 1`] = `
<div
  className="filters-main"
>
  <div>
    <Link
      className="btn"
      to="/createad"
    >
      +Ad
    </Link>
  </div>
  <form
    className="filters-form"
    onSubmit={[Function]}
  >
    <fieldset>
      <label>
        Nombre: 
      </label>
      <input
        className="input-name"
        onChange={[Function]}
        type="text"
      />
    </fieldset>
    <fieldset>
      <label>
        Operación: 
      </label>
      <select
        onChange={[Function]}
      >
        <option
          value="all"
        >
          Todas
        </option>
        <option
          value="buy"
        >
          Compra
        </option>
        <option
          value="sell"
        >
          Venta
        </option>
      </select>
    </fieldset>
    <fieldset>
      <label>
        Tag: 
      </label>
      <select
        onChange={[Function]}
      >
        <option
          value="all"
        >
          Todas
        </option>
        <option
          key="0"
          value="tag1"
        >
          tag1
        </option>
        <option
          key="1"
          value="tag2"
        >
          tag2
        </option>
        <option
          key="2"
          value="tag3"
        >
          tag3
        </option>
      </select>
    </fieldset>
    <fieldset>
      <label>
        Precio máximo: 
      </label>
      <input
        className="rangePrice"
        id="priceSelector"
        max={100}
        min="0"
        onChange={[Function]}
        type="range"
      />
      <input
        className="input-price"
        id="price"
        onChange={[Function]}
        type="text"
      />
    </fieldset>
    <fieldset>
      <input
        className="btn"
        type="submit"
      />
      <input
        className="btn"
        onClick={[Function]}
        type="reset"
      />
    </fieldset>
  </form>
</div>
`;
